cpp/dolfinx/fem/interpolate.h

```diff
@@ -835,103 +916,82 @@ void interpolate(Function<T>& u, std::span<const T> f,
   }
 }
 
-namespace impl
-{
-
-/// Interpolate from one finite element Function to another
-/// @param[in,out] u The function to interpolate into
-/// @param[in] v The function to be interpolated
-/// @param[in] cells List of cell indices to interpolate on
+/// Generate data needed to interpolate discrete functions across different
+/// meshes
+///
+/// @param[out] u The function to interpolate into
+/// @param[in] v The function to interpolate from
+/// @param[in] cells Indices of the cells in the destination mesh on which to
+/// interpolate. Should be the same as the list used when calling
+/// fem::interpolation_coords.
+/// @tparam Scalar type
 template <typename T>
-void interpolate_nonmatching_meshes(Function<T>& u, const Function<T>& v,
-                                    std::span<const std::int32_t> cells)
+NMMInterpolationData_t generate_nonmatching_meshes_interpolation_data(
+    const Function<T>& u, const Function<T>& v,
```

Comment:
I don't agree that it's an implementation detail.

1. It makes it clear to the user that nothing in the generated data is tied to the functions, only to the spaces.
2. There might be a circumstance when the user needs to make this data prior to creating a first function (I don't have a concrete example).

Reply:
Ok, I'll do it your way.